README.rst
=====================
gpsdio density plugin
=====================


.. image:: https://travis-ci.org/SkyTruth/gpsdio-density.svg?branch=master
    :target: https://travis-ci.org/SkyTruth/gpsdio-density


.. image:: https://coveralls.io/repos/SkyTruth/gpsdio-density/badge.svg?branch=master
    :target: https://coveralls.io/r/SkyTruth/gpsdio-density


A plugin for `gpsdio <https://github.com/skytruth/gpdsio/>`_ that creates density rasters.


Examples
--------

See ``gpsdio density --help`` for info.

.. code-block:: console

    $ gpsdio density tests/data/*.json example-density.tif \
        -c COMPRESS=DEFLATE \
        -c TILED=YES \
        -c PREDICTOR=2 \
        --jobs 2


Installing
----------

Via pip:

.. code-block:: console

    $ pip install gpsdio-density

From master:

.. code-block:: console

    $ git clone https://github.com/SkyTruth/gpsdio-density
    $ cd gpsdio-density
    $ pip install .


Developing
----------

.. code-block::

    $ git clone https://github.com/SkyTruth/gpsdio-density
    $ cd gpsdio-density
    $ virtualenv venv && source venv/bin/activate
    $ pip install -e .[test]
    $ py.test tests --cov gpsdio_density --cov-report term-missing


CHANGES.txt
Changes
=======

0.2 (2015-06-01)
----------------

User can now sum a field instead of just computing density - #5.
Added a `--version` flag.
Small internal performance boost.


0.1.1 (2015-05-29)
------------------

gpsdio I/O args are now passed to `gpsdio.open()` - #4
CRS info is now properly applied to the output raster - #3


0.1 (2015-05-28)
----------------

Initial release - compatible with `gpsdio>=0.0.2`.


tests/test_core.py
"""
Unittests for `gpsdio_density.core`.
"""


import tempfile

from click.testing import CliRunner
import gpsdio.cli
import rasterio as rio

import gpsdio_density


def test_standard_with_res_and_shape():

    with tempfile.NamedTemporaryFile('r+') as f:
        for res_shape in (['--res', '10'], ['--shape', '18', '36']):
            f.seek(0)
            args = [
                '--i-drv', 'NewlineJSON',
                '--i-cmp', 'GZIP',
                'density',
                'tests/data/messages1.json.gz',
                'tests/data/messages2.json.gz',
                '-c', 'COMPRESS=DEFLATE',
                '-c', 'PREDICTOR=2',
                '-c', 'ZLEVEL=9',
                '-c', 'BIGTIFF=NO',
                '-c', 'TILED=YES',
                '-c', 'INTERLEAVE=BAND',
                '--dtype', 'Int16',
                f.name
            ]
            args += res_shape
            result = CliRunner().invoke(gpsdio.cli.main.main_group, args)
            f.seek(0)

            assert result.exit_code is 0

            with rio.open(f.name) as actual,\
                    rio.open('tests/data/test-standard-expected.tif') as expected:

                assert actual.meta == expected.meta
                assert actual.read(indexes=1).all() == expected.read(indexes=1).all()

                # If the affine or transform is not properly written then
                # one of these will be incorrect.  Should have world extent.
                assert expected.bounds == actual.bounds == (-180, -90, 180, 90)


def test_with_field():

    with tempfile.NamedTemporaryFile('r+') as f:
        result = CliRunner().invoke(gpsdio.cli.main.main_group, [
            'density',
            'tests/data/messages1.json.gz',
            'tests/data/messages2.json.gz',
            '--res', '10',
            '--field', 'val',
            '--dtype', 'Int16',
            '-c', 'COMPRESS=DEFLATE',
            '-c', 'PREDICTOR=2',
            '-c', 'ZLEVEL=9',
            '-c', 'BIGTIFF=NO',
            '-c', 'TILED=YES',
            '-c', 'INTERLEAVE=BAND',
            f.name
        ])
        f.seek(0)
        print(result.output)
        assert result.exit_code is 0

        with rio.open(f.name) as actual, \
                rio.open('tests/data/test-with-field-expected.tif') as expected:

            assert actual.meta == expected.meta
            assert actual.read(indexes=1).all() * 2 == expected.read(indexes=1).all()
            assert expected.bounds == actual.bounds == (-180, -90, 180, 90)


def test_version():
    result = CliRunner().invoke(gpsdio.cli.main.main_group, [
        'density',
        '--version'
    ])

    assert result.exit_code is 0

    assert 'gpsdio-density' in result.output and gpsdio_density.__version__ in result.output


gpsdio_density/__init__.py
"""
A plugin for `gpsdio` that rasterizes positional AIS messages.
"""


__version__ = '0.1'
__author__ = 'Kevin Wurster'
__email__ = 'kevin@skytruth.org'
__source__ = 'https://github.com/SkyTruth/gpsdio-rasterize'
__license__ = """
Copyright 2015 SkyTruth

Authors: Kevin Wurster <kevin@skytruth.org>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""


from .core import rasterize


gpsdio_density/core.py
"""
Core components for gpsdio_density
"""


from __future__ import division

import logging
import math
from multiprocessing import cpu_count
from multiprocessing import Pool
import pickle

import affine
import click
import gpsdio
import numpy as np
import rasterio as rio

from . import __version__


logging.basicConfig()
log = logging.getLogger('gpsdio-density')


def _cb_key_val(ctx, param, value):

    """
    Click callback to validate and convert `--opt key=val --opt key2=val2` to
    `{'key': 'val', 'key2': 'val2'}`.

    Returns
    -------
    dict
    """

    output = {}
    for pair in value:
        if '=' not in pair:
            raise click.BadParameter("incorrect syntax for KEY=VAL argument: `%s'" % pair)
        else:
            key, val = pair.split('=')
            output[key] = val

    return output


def _cb_bbox(ctx, param, value):

    """
    Click callback to validate a bounding box.

    Returns
    -------
    tuple
        (x_min, y_min, x_max, y_max)
    """

    x_min, y_min, x_max, y_max = value
    if (x_min > y_max) or (y_min > y_max):
        raise click.BadParameter('self-intersection: {bbox}'.format(value))

    return value


def _cb_res(ctx, param, value):

    """
    Click callback to validate and transform `--res`.  Forces all values to be
    positive.  If `--res` is only given once then the value is used for both
    X and Y.

    Returns
    -------
    tuple
        Element 0 is X res and element 1 is y res.
    """

    value = tuple(map(abs, value))

    # If no values are given still return the value.  There is an error check
    # later that ensures --res OR --shape are given.
    if len(value) is 2 or len(value) is 0:
        return value
    elif len(value) is 1:
        return value[0], value[0]
    else:
        raise click.BadParameter('specified too many times.')


def _cb_shape(ctx, param, value):

    """
    Click callback to validate `--shape`.

    Returns
    -------
    tuple
        (height, width)
    """

    for v in value:
        if not v >= 1:
            raise click.BadParameter('values must be >= 1')

    return value


def _processor(args):

    """
    Create an empty numpy array, open a file, and iterate over all the messages
    with latitude and longitude.  Every time a point intersects an array element
    the element gets a +1.

    Parameters
    ----------
    args : dict
        ctx_obj : dict
            The `ctx.obj` from the parent Click context if it is a dictionary,
            or an empty dict.
        filepath : str
            Path to the file to process.
        meta : dict
            Metadata for a `rasterio` raster.  There are issues pickling
            `affine.Affine()` so `meta['transform']` are the affine elements
            as a tuple and a local instance of `affine.Affine()` is
            constructed before processing.
        field : str or None
            If summing a field rather than computing density this is the field name.

    Returns
    -------
    np.array
    """

    ctx_obj = args['ctx_obj']
    filepath = args['filepath']
    meta = args['meta']
    field = args['field']

    filter_expr = "isinstance(msg.get('lat'), (int, float)) and " \
                  "isinstance(msg.get('lon'), (int, float))"

    if field is not None:
        filter_expr += " and '%s' in msg" % field

    log.debug("Starting %s" % filepath)
    log.debug("Filter expr: %s" % filter_expr)

    data = np.zeros((meta['height'], meta['width']), dtype=meta['dtype'])
    width = meta['width']
    height = meta['height']
    aff = meta['transform']
    with gpsdio.open(filepath, driver=ctx_obj.get('i_drv'),
                     compression=ctx_obj.get('i_cmp')) as src:
        for msg in gpsdio.filter(
                src, "isinstance(msg.get('lat'), (int, float)) and "
                     "isinstance(msg.get('lon'), (int, float))"):
            col, row = (msg['lon'], msg['lat']) * ~aff
            if 0 <= row < height and 0 <= col < width:
                if field is not None:
                    val = msg[field]
                else:
                    val = 1
                data[row][col] += val

    return data


@click.command(name='density')
@click.version_option(__version__, prog_name='gpsdio-density')
@click.argument('infiles', nargs=-1, required=True)
@click.argument('outfile', required=True)
@click.option(
    '-c', '--creation-option', 'creation_options', metavar='NAME=VAL',
    callback=_cb_key_val, multiple=True,
    help='Driver-specific creation options for output raster.'
)
@click.option(
    '-f', '--format', '--driver', metavar='NAME', default='GTiff',
    help='Output format. (default: GTiff)'
)
@click.option(
    '-j', '--jobs', type=click.IntRange(1, cpu_count()), default=1,
    help='Number of files to process in parallel. (default: 1)'
)
@click.option(
    '--bbox', metavar='X_MIN Y_MIN X_MAX Y_MAX', nargs=4, type=click.FLOAT,
    default=(-180, -90, 180, 90), callback=_cb_bbox,
    help='Only process data within the specified bounding box.  '
         '(default: -180, -90, 180, 90)'
)
@click.option(
    '-n', '--nodata', type=click.FLOAT, default=None,
    help="Nodata value for output raster. (default: None)"
)
@click.option(
    '--res', metavar='RES [RES]', type=click.FLOAT, multiple=True, callback=_cb_res,
    help="Output resolution in georeferenced units.  Cannot be combined with `--shape`."
)
@click.option(
    '-s', '--shape', type=click.INT, nargs=2, metavar="ROWS COLS", callback=_cb_shape,
    help="Output shape in rows and cols.  Cannot be combined with `--res`."
)
@click.option(
    '--crs', default='EPSG:4326',
    help="Specify CRS for input points and output raster.  No transformations are performed. "
         "(default: EPSG:4326)"
)
@click.option(
    '--dtype', default=rio.int32, metavar='TYPE',
    help="Numpy datatype to use for output raster.  (default: int32)"
)
@click.option(
    '--field', metavar='NAME',
    help="Rather than computing density, sum a field.  Each point that intersects a pixel "
         "will add the value of the specified field instead of just adding 1.  Make sure "
         "to match field type with `--dtype`."
)
@click.pass_context
def compute_density(ctx, infiles, outfile, creation_options, driver, jobs, bbox,
              shape, res, crs, dtype, nodata, field):

    """
    Create a density raster from positional messages.

    Message data is streamed but the rasters are held in-memory so the max
    output resolution for the entire world is about 0.05.  Specifying a bbox
    allows for a smaller cell size but does not allow for global rasters.
    """

    x_min, y_min, x_max, y_max = bbox

    if res and not shape:
        x_res, y_res = res
        width = math.ceil((x_max - x_min) / x_res)
        height = math.ceil((y_max - y_min) / y_res)
    elif shape and not res:
        height, width = shape
        x_res = (x_max - x_min) / width
        y_res = (y_max - y_min) / height
    else:
        raise click.BadParameter('must specify `--res` OR `--shape`')

    meta = {
        'driver': driver,
        'height': height,
        'width': width,
        'transform': affine.Affine(x_res, 0.0, x_min, 0.0, -y_res, y_max),
        'crs': crs,
        'dtype': dtype,
        'nodata': nodata,
        'count': 1
    }
    meta.update(**creation_options)

    # Click's `ctx` object is an instance of `click.Context()` which cannot immediately
    # be pickled, which means it cannot be directly passed to the subprocess.
    # The following does the following:
    #
    #   1. If `ctx.obj` is a dict that is pickleable, pass it as is
    #   2. If `ctx.obj` is a dict that IS NOT picklable, just pass an empty dict
    #   3. If `ctx.obj` is NOT a dict, just pass an empty dict.
    #
    # These checks are required because the user may be attaching this plugin
    # to some CLI that doesn't behave like gpsdio.
    if isinstance(ctx.obj, dict):
        ctx_obj = ctx.obj.copy()
        try:
            pickle.dumps(ctx_obj)
        except pickle.PicklingError:
            log.warning("Parent click `ctx.obj` cannot be pickled - passing an empty dict.  "
                        "Some functionality may not be available.")
            ctx_obj = {}
    else:
        log.warning("Parent click `ctx.obj` is not a dictionary.  Some functionality may not "
                    "be available")
        ctx_obj = {}

    task_generator = (
        {
            'ctx_obj': ctx_obj,
            'filepath': fp,
            'meta': meta,
            'field': field
        } for fp in infiles)

    with rio.open(outfile, 'w', **meta) as dst:

        output = sum(
            (a for a in Pool(jobs).imap_unordered(_processor, task_generator)))

        dst.write(output.astype(dst.meta['dtype']), indexes=1)


if __name__ == '__main__':
    compute_density()
